From: YAOF-1 <dev@localhost>
Date: Thu, 1 Jan 1970 00:00:00 +0000
Subject: [PATCH] libnftnl: add brcm fullconenat support

Adds the masquerade address-range registers that the brcmfullcone nft
statement uses to mark a full-cone masquerade rule.

Regenerated against the libnftnl version OpenWrt pins (1.3.1); the previous
version of this change was written against an older libnftnl and applied with
offsets, which put its hunks at the wrong place.

Self-contained: it does not depend on the nftables "fullcone" expression
support that was dropped together with nft-fullcone.

Signed-off-by: YAOF-1 <dev@localhost>
---
--- a/include/libnftnl/expr.h
+++ b/include/libnftnl/expr.h
@@ -269,6 +269,8 @@
 	NFTNL_EXPR_MASQ_FLAGS		= NFTNL_EXPR_BASE,
 	NFTNL_EXPR_MASQ_REG_PROTO_MIN,
 	NFTNL_EXPR_MASQ_REG_PROTO_MAX,
+	NFTNL_EXPR_MASQ_REG_ADDR_MIN,
+	NFTNL_EXPR_MASQ_REG_ADDR_MAX,
 	__NFTNL_EXPR_MASQ_MAX
 };
 
--- a/include/linux/netfilter/nf_tables.h
+++ b/include/linux/netfilter/nf_tables.h
@@ -1499,12 +1499,16 @@
  * @NFTA_MASQ_FLAGS: NAT flags (see NF_NAT_RANGE_* in linux/netfilter/nf_nat.h) (NLA_U32)
  * @NFTA_MASQ_REG_PROTO_MIN: source register of proto range start (NLA_U32: nft_registers)
  * @NFTA_MASQ_REG_PROTO_MAX: source register of proto range end (NLA_U32: nft_registers)
+ * @NFTA_MASQ_REG_ADDR_MIN: source register of address range start (NLA_U32: nft_registers)
+ * @NFTA_MASQ_REG_ADDR_MAX: source register of address range end (NLA_U32: nft_registers)
  */
 enum nft_masq_attributes {
 	NFTA_MASQ_UNSPEC,
 	NFTA_MASQ_FLAGS,
 	NFTA_MASQ_REG_PROTO_MIN,
 	NFTA_MASQ_REG_PROTO_MAX,
+	NFTA_MASQ_REG_ADDR_MIN,
+	NFTA_MASQ_REG_ADDR_MAX,
 	__NFTA_MASQ_MAX
 };
 #define NFTA_MASQ_MAX		(__NFTA_MASQ_MAX - 1)
--- a/src/expr/masq.c
+++ b/src/expr/masq.c
@@ -20,6 +20,8 @@
 	uint32_t		flags;
 	enum nft_registers	sreg_proto_min;
 	enum nft_registers	sreg_proto_max;
+	enum nft_registers	sreg_addr_min;
+	enum nft_registers	sreg_addr_max;
 };
 
 static int
@@ -38,6 +40,12 @@
 	case NFTNL_EXPR_MASQ_REG_PROTO_MAX:
 		memcpy(&masq->sreg_proto_max, data, data_len);
 		break;
+	case NFTNL_EXPR_MASQ_REG_ADDR_MIN:
+		memcpy(&masq->sreg_addr_min, data, data_len);
+		break;
+	case NFTNL_EXPR_MASQ_REG_ADDR_MAX:
+		memcpy(&masq->sreg_addr_max, data, data_len);
+		break;
 	}
 	return 0;
 }
@@ -58,6 +66,12 @@
 	case NFTNL_EXPR_MASQ_REG_PROTO_MAX:
 		*data_len = sizeof(masq->sreg_proto_max);
 		return &masq->sreg_proto_max;
+	case NFTNL_EXPR_MASQ_REG_ADDR_MIN:
+		*data_len = sizeof(masq->sreg_addr_min);
+		return &masq->sreg_addr_min;
+	case NFTNL_EXPR_MASQ_REG_ADDR_MAX:
+		*data_len = sizeof(masq->sreg_addr_max);
+		return &masq->sreg_addr_max;
 	}
 	return NULL;
 }
@@ -73,6 +87,8 @@
 	switch (type) {
 	case NFTA_MASQ_REG_PROTO_MIN:
 	case NFTA_MASQ_REG_PROTO_MAX:
+	case NFTA_MASQ_REG_ADDR_MIN:
+	case NFTA_MASQ_REG_ADDR_MAX:
 	case NFTA_MASQ_FLAGS:
 		if (mnl_attr_validate(attr, MNL_TYPE_U32) < 0)
 			abi_breakage();
@@ -96,6 +112,12 @@
 	if (e->flags & (1 << NFTNL_EXPR_MASQ_REG_PROTO_MAX))
 		mnl_attr_put_u32(nlh, NFTA_MASQ_REG_PROTO_MAX,
 				 htobe32(masq->sreg_proto_max));
+	if (e->flags & (1 << NFTNL_EXPR_MASQ_REG_ADDR_MIN))
+		mnl_attr_put_u32(nlh, NFTA_MASQ_REG_ADDR_MIN,
+				 htonl(masq->sreg_addr_min));
+	if (e->flags & (1 << NFTNL_EXPR_MASQ_REG_ADDR_MAX))
+		mnl_attr_put_u32(nlh, NFTA_MASQ_REG_ADDR_MAX,
+				 htonl(masq->sreg_addr_max));
 }
 
 static int
@@ -121,6 +143,16 @@
 			be32toh(mnl_attr_get_u32(tb[NFTA_MASQ_REG_PROTO_MAX]));
 		e->flags |= (1 << NFTNL_EXPR_MASQ_REG_PROTO_MAX);
 	}
+	if (tb[NFTA_MASQ_REG_ADDR_MIN]) {
+		masq->sreg_addr_min =
+			ntohl(mnl_attr_get_u32(tb[NFTA_MASQ_REG_ADDR_MIN]));
+		e->flags |= (1 << NFTNL_EXPR_MASQ_REG_ADDR_MIN);
+	}
+	if (tb[NFTA_MASQ_REG_ADDR_MAX]) {
+		masq->sreg_addr_max =
+			ntohl(mnl_attr_get_u32(tb[NFTA_MASQ_REG_ADDR_MAX]));
+		e->flags |= (1 << NFTNL_EXPR_MASQ_REG_ADDR_MAX);
+	}
 
 	return 0;
 }
@@ -145,6 +177,16 @@
 		ret = snprintf(buf + offset, remain, "flags 0x%x ", masq->flags);
 		SNPRINTF_BUFFER_SIZE(ret, remain, offset);
 	}
+	if (e->flags & (1 << NFTNL_EXPR_MASQ_REG_ADDR_MIN)) {
+		ret = snprintf(buf + offset, remain,
+			       "addr_min reg %u ", masq->sreg_addr_min);
+		SNPRINTF_BUFFER_SIZE(ret, remain, offset);
+	}
+	if (e->flags & (1 << NFTNL_EXPR_MASQ_REG_ADDR_MAX)) {
+		ret = snprintf(buf + offset, remain,
+			       "addr_max reg %u ", masq->sreg_addr_max);
+		SNPRINTF_BUFFER_SIZE(ret, remain, offset);
+	}
 
 	return offset;
 }
